technical-docs/k8ssandra-upfront-validations.md

```diff
@@ -0,0 +1,606 @@
+# K8ssandra upfront validations
+Author: @jeffbanks
+
+## Enhancement
+Request for validator tools to be used to notify of invalid K8ssandra configurations.  Rules will be customizable and support detection of issues earlier in the deployment process.
+
+For example, in the K8ssandra jvm-option heap settings there is a requirement to support the following property structure when customized.
+
+```yaml
+cassandra:
+  # cluster-level settings
+  heap:
+    # min & max equal
+    size: 700M
+    newGenSize: 1.4Gi
+  # dc specific settings
+  dataCenters:
+    - name: dc1
+      size: 1
+      heap:
+        size: 600M
+        newGenSize: 1.2Gi
+    - name: dc2
+      size: 1
+```
+
+In the example, when values are not specified, a heap property is applied without any child properties.  These types of issues could be customized and flagged as invalid settings, based on K8ssandra best-practices or customized as needed.   
```

Comment:
That is certainly an option and we have some validation logic embedded in our template helpers at current. From my experience on a few of these template changes, it starts to become cumbersome to maintain both the `template logic` and the `validation logic`.  Even though the sprig functions and go templating provide features for customization of logic and validations, it has become a bit messy and error-prone IMO.  

For example in [this template helper](https://github.com/k8ssandra/k8ssandra/blob/main/charts/k8ssandra/templates/_helpers.tpl#L1) we have a good [example](https://github.com/k8ssandra/k8ssandra/blob/87cbf6f3ca359c45fb46901717b4d812b14c81ed/charts/k8ssandra/templates/_helpers.tpl#L255) of having to address formatting, validation, and mapping of values.  For our failure conditions, we don't get a collection of issues that may have occurred.  It would be nice to provide a full report of all the issues found for the values being applied.

The tools proposed, focus solely on the overall rendered output (that could be generated using template helpers) w/ the validation decoupled from the effort of formatting and mapping of values.

For the operator based implementation, I like the use of marker comments in `kubebuilder` to support schema validation.  Where I still see a gap, is the overall application of k8s best-practices and security validations that can be reported based on specific k8s versions.

Question: @bradfordcp Maybe the use of the upfront validation tools proposed would better be part of a k8ssandra UI or a CLI that can inspect k8ssandra configurations prior to final apply/deployment? That is, an optional tool for the community to use.   

It seems reasonable to have one or more static analysis tools along with schema validators.  Where schema validations would be applied using something like `kubebuilder` markers.

Example:
```go 
type MySpec struct {
    // +kubebuilder:validation:MaxLength=30
    // +kubebuilder:validation:MinLength=1
    Name string `json:"name,omitempty"`

    // +kubebuilder:validation:MaxItems=20
    // +kubebuilder:validation:MinItems=1
    // +kubebuilder:validation:UniqueItems=false
    Tickets[]string `json:"tickets,omitempty"`  
}
```